text/0000-dockerfiles.md

```diff
@@ -0,0 +1,209 @@
+The `build_id` arg allows the Dockerfile to invalidate the cache after a certain layer and must be defaulted to `0`.
+When the `$build_id` arg is referenced in a `RUN` instruction, all subsequent layerrs will be rebuilt on the next build (as the value will change).
+
+A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of builder-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before a runtime Dockerfile is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.
+
+Finally, base images may be statically labeled with any number of `provides` that are treated as build plan entries.
+These `provides` may contain fields other than `name`, which, when mismatched with `requires`, mark the entry as `unmet` by the stack.
+This is important to ensure that:
+- Rebasing always remains an option for end users.
+- Buildpacks do not become dependent on hooks.
+- Builds can be time-optimized by creating base images ahead of time.
+
+NOTE: the above can be accomplished by a hook with a no-op `/bin/build` -- do we really need this?
+
+#### Example: App-specified Dockerfile Hook
+
+This example hook would allow an app to provide runtime and build-time base image extensions as "run.Dockerfile" and "build.Dockerfile."
+The app developer can decide whether the extensions are rebasable.
+
+##### `/cnb/hooks/com.example.apphook/bin/build`
+```
+#!/bin/sh
+[ -f build.Dockerfile ] && cp build.Dockerfile "$1/"
+[ -f run.Dockerfile ] && cp run.Dockerfile "$1/"
+```
+
+#### Example: RPM Dockerfile Hook (app-based)
+
+This example hook would allow a builder to install RPMs for each language runtime, based on the app directory.
+
+Note: The Dockerfiles referenced must disable rebasing, and build times will be slower compared to buildpack-provided runtimes.
+
+##### `/cnb/hooks/com.example.rpmhook/bin/build`
+```
+#!/bin/sh
+[ -f Gemfile.lock ] && cp "$BUILDPACK_DIR/Dockerfile-ruby" "$1/Dockerfile"
+[ -f package.json ] && cp "$BUILDPACK_DIR/Dockerfile-node" "$1/Dockerfile"
+```
+
+
+### Dockerfiles for Base Images
+
+The same Dockerfile format may be used to create new base images or modify existing base images outside of the app build process (e.g., before creating a builder). Any specified labels override existing values.
+
+Dockerfiles that are used to create a base image must create a `/cnb/image/genpkgs` executable that outputs a [CycloneDX](https://cyclonedx.org)-formatted list of packages in the image with PURL IDs when invoked. This executable is executed after all Dockerfiles are applied, and the output replaces the label `io.buildpacks.sbom`. This label doubles as a Software Bill-of-Materials for the base image. In the future, this label will serve as a starting point for the application SBoM.
```

Comment:
Can we just use the `build`/`extend` script to output this instead?

Reply:
The original SBoM needs to be replaced after additional OS packages are installed, so both mechanisms are needed.

(also answered during WG)

Reply:
/queue-issue buildpacks/lifecycle "Lifecycle run /cnb/image/genpkgs on the extended run image" epic/dockerfiles type/enhancement

Reply:
Marking as resolved given that we're leaving SBOM generation up to the platform.